packages/ai-jsx/src/lib/openai.tsx

```diff
@@ -291,3 +295,51 @@ export async function* OpenAIChatModel(
 
   return currentMessage.content;
 }
+
+/**
+ * Generates an image from a prompt using the DALL-E model.
+ *
+ * @returns the URL of the generated image.
+ *          If num_samples is greater than 1, URLs are separated by newlines.
+ */
+export async function DalleImageGen(
+  { num_samples = 1, size = '512x512', children }: ImageGenPropsWithChildren,
+  { render, getContext, logger }: LLMx.ComponentContext
+) {
+  let prompt = await render(children);
+  const openai = getContext(openAiClientContext);
+
+  let sizeEnum;
+  switch (size) {
+    case '256x256':
+      sizeEnum = CreateImageRequestSizeEnum._256x256;
+    case '512x512':
+      sizeEnum = CreateImageRequestSizeEnum._512x512;
+    case '1024x1024':
+      sizeEnum = CreateImageRequestSizeEnum._1024x1024;
+  }
+
+  if (sizeEnum == undefined) {
+    throw new Error(`Invalid size ${size}. Dalle only supports 256x256, 512x512, and 1024x1024`);
+  }
+
+  const imageRequest = {
+    prompt: prompt,
+    n: num_samples,
+    size: sizeEnum,
+    response_format: CreateImageRequestResponseFormatEnum.Url,
+  };
+
+  logger.debug({ imageRequest }, 'Calling createImage');
+
+  const response = await openai.createImage(imageRequest);
+
+  if (response.status < 200 || response.status >= 300) {
+    throw new Error(`createImage failed with status ${response.status}`);
+  } else {
+    logger.debug({ statusCode: response.status, headers: response.headers }, 'createImage succeeded');
+  }
```

Comment:
I originally wanted to use `checkOpenAIResponse`, but since this is not streaming, that might not be a good idea. I was hesitant about adding "createImage" to `OpenAIMethod`, but if I do I can throw an `OpenAIError` instead.

Reply:
I think `OpenAIError` is what we want here.

Ultimately, we want to provide a layer on top of OpenAI to make the errors nicer – for instance, explaining that 429 is a rate limit, etc. If the OpenAI APIs are consistent across DallE / GPT, etc, then it would be nice for our checking to be consistent too.

Reply:
I think `OpenAIError` is what we want here.

Ultimately, we want to provide a layer on top of OpenAI to make the errors nicer – for instance, explaining that 429 is a rate limit, etc. If the OpenAI APIs are consistent across DallE / GPT, etc, then it would be nice for our checking to be consistent too.